### PR TITLE
Add option to prevent App Nap.

### LINKED ISSUE #8883: Task.Delay takes too long on macOS
**Describe the bug**
In an Avalonia macOS desktop app Task.Delay often takes too long to return.
In a simple non-Avalonia console app the issue does not occur.

**To Reproduce**

1. Create a minimal Avalonia app from the VS project template.
2. Somewhere (e.g. MainWindow) put Task.Delay in an infinite loop.
3. Often Task.Delay takes way longer than configured.

**Expected behavior**
Task.Delay takes about as long as configured.

**Desktop (please complete the following information):**
- OS: Mac
- Version: 12.5

**Additional context**
Repro: https://github.com/Stannieman/AvaloniaDelayTest
This loops a Task.Delay of 250ms. When delay takes longer than 500ms it prints the actual duration to the terminal.

After only a couple of minutes the output looks like this:
```
Too long: 00:00:10.1278680
Too long: 00:00:01.4919020
Too long: 00:00:10.1362750
Too long: 00:00:10.0013450
Too long: 00:00:10.0012170
Too long: 00:00:10.0014200
Too long: 00:00:10.0012190
Too long: 00:00:10.0014000
Too long: 00:00:10.0014770
Too long: 00:00:10.0012340
Too long: 00:00:10.0298920
Too long: 00:00:09.9944410
Too long: 00:00:10.0013260
Too long: 00:00:10.0011620
Too long: 00:00:10.0011870
Too long: 00:00:10.0100230
Too long: 00:00:10.0057740
Too long: 00:00:10.0125380
Too long: 00:00:10.0072980
Too long: 00:00:10.0296070
Too long: 00:00:09.9992390
Too long: 00:00:10.0023670
Too long: 00:00:10.0020110
Too long: 00:00:10.0019850
Too long: 00:00:10.0063740
Too long: 00:00:09.9987900
Too long: 00:00:10.0016140
Too long: 00:00:01.4810180
Too long: 00:00:10.0016090
Too long: 00:00:20.0312540
Too long: 00:00:19.9925110
Too long: 00:00:20.0018100
Too long: 00:00:20.0017170
Too long: 00:00:20.0116750
Too long: 00:00:19.9987460
Too long: 00:00:20.0019690
Too long: 00:00:20.0016450
Too long: 00:00:20.0018110
Too long: 00:00:03.6547710
Too long: 00:00:02.4756200
Too long: 00:00:02.5731120
Too long: 00:00:03.6644760
```

Since Task.Delay behind the scenes uses the same timer mechanism as most if not all other .NET timer APIs this is quite concerning.

Is Avalonia messing with doing something thread pool/scheduling stuff somewhere? The machine is idle when this occurs. Memory is not full, no CPU spikes. It just "happens".

### PR DESCRIPTION
**Pull request by Stannieman**
## What does the pull request do?
Add an option to prevent App Nap. This is a macOS feature that can suspend apps that are not in focus. Apps that are performing time critical tasks such as periodically doing something in the background are affected by this.

## What is the current behavior?
Nothing stops macOS from suspending the app.

## What is the updated/expected behavior with this PR?
The Avalonia.Platform.PlatformBehavior.RequestPlatformBehaviorInhibition method can be called to disable App Nap.

## How was the solution implemented (if it's not obvious)?
Add a new API that can be used to (temporarily) disable platform specific behavior.

## Checklist

- [ ] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes
None

## Obsoletions / Deprecations
None

## Fixed issues
Fixes #8883

### CODE REVIEW

File: native/Avalonia.Native/src/OSX/main.mm
Comment (Stannieman) [inline native/Avalonia.Native/src/OSX/main.mm:150]: Something is wrong here.
When I assign to the "static" (in anonymous namespace here but that is equivalent) `s_preventAppNapHandle` I get a `DllNotFoundException` for `libAvaloniaNative` when the app starts. When `s_preventAppNapHandle` is a local variable this is not the case, and when I do not assign to `s_preventAppNapHandle` but still call `beginActivityWithOptions` then there it also loads.

Making `s_preventAppNapHandle` a static class variable has the same problem.

The thing is that `s_preventAppNapHandle` needs to be something static or in a singleton because as soon as this object is deallicated App Nap will turn on again.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0023557-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: native/Avalonia.Native/src/OSX/main.mm
Comment (Stannieman) [inline native/Avalonia.Native/src/OSX/main.mm:150]: We need to keep a reference to the result of `beginActivityWithOptions`. App Nap will turn on again when this is deallocated.

For this reasom `s_preventAppNapHandle` is in an anomymous namespace at the tom (equivalent to static).

Question: Do we need thread safety here? Or should this only be called from main thread anyway? What is the policy?

File: src/Avalonia.Native/AvaloniaNativePlatform.cs
Comment (Stannieman) [inline src/Avalonia.Native/AvaloniaNativePlatform.cs:194]: How do I test this?

**Comment by Stannieman** [comment]
> You can test this PR using the following package version. `11.0.999-cibuild0023557-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

Just want to say that this is super convenient for if you want to test your change in an actual app! No need to copy the source code into your own codebase or mimicing the CI to produce a NuGet package on your own. Big up!

**Comment by kekekeks** [comment]
I don't think it should be am app-wide setting, that would incentive invalid app behavior. The app is supposed to ask the OS to not use app-nap while the app is doing a critical operation, not all the time the app is running.

**Comment by Stannieman** [comment]
> I don't think it should be am app-wide setting, that would incentive invalid app behavior. The app is supposed to ask the OS to not use app-nap while the app is doing a critical operation, not all the time the app is running.

True. In that case the option can be removed from `MacOSPlatformOptions`. It should still be possible to toggle it whenever you want using `AvaloniaNativeMacOptions`, but do you happen to know how to use it?

I searched for `AvaloniaNativeMacOptions` and it has no usages or documentation and also AvaloniaLocator.Current does not resolve it (because it is not registered). Does it need to be registered in `AvaloniaNativePlatform.DoInitialize`? Why does this class exist if there are 0 references to it?

**Comment by kekekeks** [comment]
Note that platform options are supposed to be immutable once you've passed them to app builder. This feature requires a separate appnap API, preferably exposed via  toplevel. We can then build feature like device sleep prevention or screen highlight turnoff prevention on top of the same API.

**Comment by kekekeks** [comment]
Something like:
```cs
enum PlatformInhibitionType
{
    AppSleep,
    SystemIdle // that one will be implemented later
}
```


```diff
class TopLevel
{
+ Task<IDisposable> RequestPlatformInhibition(PlatformInhibitionType type, string reason);
}
```

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0023566-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by Stannieman** [comment]
Ok I made something that works like you suggested.

File: src/Avalonia.Base/Platform/PlatformBehavior.cs
Comment (Stannieman) [inline src/Avalonia.Base/Platform/PlatformBehavior.cs:20]: Null check here because IPlatformBehaviorInhibition is only registered in `AvaloniaNativePlatform`. Is this ok or should I right away implement it for all platforms? Also is `AvaloniaNativePlatform` actually macOS specific?

File: src/Avalonia.Native/AvaloniaNativePlatform.cs
Comment (Stannieman) [inline src/Avalonia.Native/AvaloniaNativePlatform.cs:171]: This is not used and also changing ShowInDock does not work after the app is visible anyway.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0023569-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0023576-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0023578-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by ahopper** [comment]
I'm not sure if this is useful but the following works on windows, there are further options as to exactly what is allowed such as keeping the screen on

```
public class PowerSaving
  {
      // from https://stackoverflow.com/questions/28751192/how-can-i-prevent-windows-os-from-entering-sleep-mode-net
      public enum EXECUTION_STATE : uint
      {
          ES_SYSTEM_REQUIRED = 0x00000001,
          ES_AWAYMODE_REQUIRED = 0x00000040,
          ES_CONTINUOUS = 0x80000000,
          ES_DISPLAY_REQUIRED = 0x00000002,
      }

      internal class NativeMethods
      {
          [DllImport("kernel32.dll", CharSet = CharSet.Auto, SetLastError = true)]
          public static extern EXECUTION_STATE SetThreadExecutionState(EXECUTION_STATE esFlags);
      }
      public static void PreventSleep()
      {
          NativeMethods.SetThreadExecutionState(EXECUTION_STATE.ES_CONTINUOUS | EXECUTION_STATE.ES_SYSTEM_REQUIRED );
      }
      public static void EnableSleep()
      {
           NativeMethods.SetThreadExecutionState(EXECUTION_STATE.ES_CONTINUOUS);
      }
  }
```

**Comment by Stannieman** [comment]
It looks like the options for Windows are to prevent the entire system to go to sleep. This is not equivalent to App Nap which puts this specific app to sleep only.

**Comment by kekekeks** [comment]
Inhibiting sleep and screen power saving would be a different `PlatformBehaviorInhibitionType`, yes.

**Comment by Stannieman** [comment]
How do we move forward with this? IMO there are currently 4 possibilities:
* It is OK and should be approved and merged.
* It is not OK, in which case I'd like to know what is required to make it OK.
* Handling App Nap is not Avalonia's job (which would suck it is what it is?)
* We don't know, in which case the main maintainers together should probably make a formal decision about the design. So a timeslot for making this decision should be scheduled and the final decision should be communicated.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0024280-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by StefAviator** [comment]
Can this be reviewed? We are in need of some kind of solution to App Nap on MacOS as it causes unstability for some applications.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029607-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

File: src/Avalonia.Base/Platform/PlatformBehavior.cs
Comment (maxkatz6) [inline src/Avalonia.Base/Platform/PlatformBehavior.cs:22]: Can you change the way how this interface is registered and used.
In latest preview we have introduced a new pattern how we want to use platform specific features instead of global service locator. As well as avoid static APIs in general.

ITopLevelImpl has now TryGetFeature method.
Then you can add 
```csharp
Task<IDisposable> RequestPlatformInhibition(PlatformInhibitionType type, string reason);
```
to the TopLevel, which internally will attempt to resolve platform implementation using `PlatformImpl.TryGetFeature<IPlatformBehaviorInhibition>()`.

**Comment by maxkatz6** [comment]
@Stannieman hi! I will review this PR from now. From my side there is one API problem which I commented. Aside from that it looks good.
In general, I believe that this problem shouldn't be solved by Avalonia, but we don't have much choice here, so we are going to accept it.

If you prefer, or if you are currently busy, I can complete this PR as per my review comment above and merge it after. Let me know.

File: src/Avalonia.Base/Platform/PlatformBehavior.cs
Comment (maxkatz6) [reply src/Avalonia.Base/Platform/PlatformBehavior.cs:22]: Also, new API needs XML documentation.

File: src/Avalonia.Base/Platform/PlatformBehavior.cs
Comment (maxkatz6) [reply src/Avalonia.Base/Platform/PlatformBehavior.cs:32]: Current solution with returning dummy disposable is good. API-wise there is not much difference.

**Comment by Stannieman** [comment]
> @Stannieman hi! I will review this PR from now. From my side there is one API problem which I commented. Aside from that it looks good. In general, I believe that this problem shouldn't be solved by Avalonia, but we don't have much choice here, so we are going to accept it.
> 
> If you prefer, or if you are currently busy, I can complete this PR as per my review comment above and merge it after. Let me know.

I will adjust this today according to remark.
Also I fully agree that this is not really Avalonia's job. Ideally there should be some "macOS companion" library for .NET, but AFAIK this currently exists only as part of Xamarin.

File: src/Avalonia.Base/ExecuteActionDisposer.cs
Comment (maxkatz6) [inline src/Avalonia.Base/ExecuteActionDisposer.cs:5]: Should be internal.
Or, for the Action you can use `Disposable.Create()` from the Avalonia.Reactive namespace.

File: src/Avalonia.Native/WindowImplBase.cs
Comment (maxkatz6) [inline src/Avalonia.Native/WindowImplBase.cs:516]: I think there is a misunderstanding.
TryGetFeature correctly returns _platformBehaviorInhibition, as it should.

But `Task<IDisposable> RequestPlatformInhibition` should be moved to the TopLevel control. As it's not platform specific. And stubs can be removed from other backends.

Comment (maxkatz6) [reply src/Avalonia.Native/WindowImplBase.cs:516]: I.e., when TryGetFeature returns correct IPlatformBehaviorInhibition only on macos backend (as for now), it will just return null on other backends, falling back to the `return new ExecuteActionDisposer` route in the TopLevel code (or `Disposable.Create`).

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029884-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Review by maxkatz6 — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029895-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Pull request merged**
